pkg/api/frontendsettings.go

```diff
@@ -23,6 +27,61 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
+// Returns a file that is easy to check for changes
+// Any changes to the file means we should refresh the frontend
+func (hs *HTTPServer) GetFrontendAssets(c *contextmodel.ReqContext) {
+	hash := sha256.New()
+	keys := map[string]any{}
+
+	// BuildVersion
+	hash.Reset()
+	_, _ = hash.Write([]byte(setting.BuildVersion))
+	_, _ = hash.Write([]byte(setting.BuildCommit))
+	_, _ = hash.Write([]byte(fmt.Sprintf("%d", setting.BuildStamp)))
+	keys["version"] = fmt.Sprintf("%x", hash.Sum(nil))
+
+	// Plugin configs
+	plugins := []string{}
+	for _, p := range hs.pluginStore.Plugins(c.Req.Context()) {
+		plugins = append(plugins, fmt.Sprintf("%s@%s", p.Name, p.Info.Version))
+	}
+	keys["plugins"] = sortedHash(plugins, hash)
```

Comment:
This can cause a race condition when plugins CDN is enabled.

In Grafana Enterprise, plugins that are served from CDN are loaded in the background by the `cdn.Service` background service:

https://github.com/grafana/grafana-enterprise/blob/b21ea08f821090e2c6b2f05188804463f5109af8/src/pkg/extensions/pluginsintegration/cdn/cdn.go#L27-L53

If we calculate the plugins hash here and CDN plugins haven't been loaded yet (which can be the case if the API call comes from a live connection state change in the frontend, following a backend restart), we will return the "wrong" value for the plugins hash, because plugins are still being loaded in the background.

This can cause refreshes that aren't needed in the frontend when the Grafana backend restarts. I'm not sure if this is a big deal but I wanted to point this out.

@wbrowne we may want to revisit that so CDN plugins are loaded synchronously, before the API starts responding, or something else.

cc @ryantxu

Reply:
I think if at all possible, we wouldn't want to choose to negatively impact startup time instead of invoking a page refresh. As we move more to split things, the sooner we get used to async the better 🤷

Reply:
I don't see any issue here, but @xnyo you're worried about the case where if the frontend settings are generated _before_ the CDN plugins are registered?

Reply:
Yes that's what I am worried about. It could be a problem (it already is, this PR might just increase the chances of this condition happening) but I have noticed that the reload is triggered when changing location, so it should be safe to keep everything as-is.

I'm happy to get this merged :+1:

Reply:
Thankfully not so much a problem now anyway as we have the redirect that helps us out!